### PR TITLE
Yi-VL Model

### PR DESCRIPTION
**Pull request by BabyChouSr**
Adding support for the Yi-VL Model: https://huggingface.co/01-ai/Yi-VL-6B

Note, since the original repo does not have a very friendly format, I moved the files and created my own config which makes it more compatible with the SGLang codebase. This allows us to load the model, tokenizer, and processor without much code change.

To test, simply call:
```
runtime = sgl.Runtime(model_path="BabyChou/Yi-VL-6B",
                      tokenizer_path="BabyChou/Yi-VL-6B")
```

Link to huggingface repo compatible with this commit:
6B model: https://huggingface.co/BabyChou/Yi-VL-6B
34B model: https://huggingface.co/BabyChou/Yi-VL-34B

### CODE REVIEW

**Comment by exceedzhang** [comment]
@BabyChouSr  I ran python test_openai_server.py --test-image,
test_chat_completion_image(args) error, following error:
image
image
Server Error:
image

File: python/sglang/srt/utils.py [hunk 1]
```diff
@@ -191,11 +191,11 @@ def ret_func(grid, num_warps, *args):
 
 def is_multimodal_model(model):
     if isinstance(model, str):
-        return "llava" in model
+        return "llava" or "yi-vl" in model
     from sglang.srt.model_config import ModelConfig
 
     if isinstance(model, ModelConfig):
-        return "llava" in model.path.lower()
+        return "llava" in model.path.lower() or "yi-vl" in model.path.lower()
```
Comment (merrymercy) [inline python/sglang/srt/utils.py]:
```suggestion
        model_path = model.path.lower()
        return "llava" in model_path or "yi-vl" in model_path
```

**Review by merrymercy — COMMENTED** [review]
Looks good to me!

1. Could you also convert the 34B version?
2. Could you upload the conversion script?
3. Could you add an example similar to this one https://github.com/sgl-project/sglang/blob/main/examples/quick_start/srt_example_llava.py, but for Yi-VL?

**Comment by paulcx** [comment]
try an example in srt_example_llava.py

state = image_qa.run( image_path="images/cat.jpeg", question="What is this?", max_new_tokens=64) then

```
    File "xx/sglang/python/sglang/srt/models/llava.py", line 63, in pad_input_ids
        offset = input_ids.index(self.config.image_token_index)
ValueError: 64002 is not in list
```

**Comment by aliozts** [comment]
@paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to https://github.com/sgl-project/sglang/issues/41

**Comment by paulcx** [comment]
> @paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to https://github.com/sgl-project/sglang/issues/41

Should it be <image_holder>?

**Comment by aliozts** [comment]
I don't know the placeholder for Yi models. it can be different but the key part is the respective token being missing most probably since the error is caused by `input_ids.index(self.config.image_token_index)` means that the respective token(64002)'s text version is missing.

**Comment by BabyChouSr** [comment]
> > @paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to #41
> 
> Should it be <image_holder>?

I believe that the sglang frontend language (using the example from `srt_example_llava` but changing llava for yi-vl model) will automatically add the image token which is `<image_placeholder>`

**Comment by paulcx** [comment]
> > > @paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to #41
> 
> > 
> 
> > Should it be <image_holder>?
> 
> 
> 
> I believe that the sglang frontend language (using the example from `srt_example_llava` but changing llava for yi-vl model) will automatically add the image token which is `<image_placeholder>`

If so, why does "64002 is not in list" happen?

**Comment by BabyChouSr** [comment]
> > > > @paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to #41
> > 
> > 
> > > 
> > 
> > 
> > > Should it be <image_holder>?
> > 
> > 
> > I believe that the sglang frontend language (using the example from `srt_example_llava` but changing llava for yi-vl model) will automatically add the image token which is `<image_placeholder>`
> 
> If so, why does "64002 is not in list" happen?

Are you using the following for your runtime?

```
runtime = sgl.Runtime(model_path="BabyChou/Yi-VL-6B",
                      tokenizer_path="BabyChou/Yi-VL-6B")
```

**Comment by paulcx** [comment]
> > > > > @paulcx I had the same issue, it's because of the prompting, you need to include the `<image>` from my understanding. I'd suggest referring to #41
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Should it be <image_holder>?
> 
> > > 
> 
> > > 
> 
> > > I believe that the sglang frontend language (using the example from `srt_example_llava` but changing llava for yi-vl model) will automatically add the image token which is `<image_placeholder>`
> 
> > 
> 
> > If so, why does "64002 is not in list" happen?
> 
> 
> 
> Are you using the following for your runtime?
> 
> 
> 
> ```
> 
> runtime = sgl.Runtime(model_path="BabyChou/Yi-VL-6B",
> 
>                       tokenizer_path="BabyChou/Yi-VL-6B")
> 
> ```

I used the runtime of endpoint. It failed with same error using cli above.

**Comment by BabyChouSr** [comment]
@paulcx I posted some new changes, try running:

```
python3 srt_example_yi_vl.py
```

**Comment by paulcx** [comment]
> @paulcx I posted some new changes, try running:
> 
> 
> 
> ```
> 
> python3 srt_example_yi_vl.py
> 
> ```

will try later. btw i found yi-vl architecture and model class were not found in model runner by hf config only. so i changed with hard code for now.

**Comment by paulcx** [comment]
> @paulcx I posted some new changes, try running:
> 
> ```
> python3 srt_example_yi_vl.py
> ```

It works.

Also works with:
```
curl http://localhost/generate -H "Content-Type: application/json" -d '{"text": "### Human: <image_placeholder>\n图片里有什么？\n### Assistant:", "image_data": "xxxx", "sampling_params": {"max_new_tokens": 64, "temperature": 0, "stop": "### "}}'
```

**Review by merrymercy — APPROVED** [review]
@BabyChouSr Ready to be merged?

**Comment by BabyChouSr** [comment]
> @BabyChouSr Ready to be merged?

Yup!

**Comment by loveunk** [comment]
It looks like this PR has been merged to SGLang v0.1.11.
However, still encountered the `ValueError: 64002 is not in list` issue while performing inference with `llava-1.6-34B` and SGLang v0.1.11.

**Comment by BabyChouSr** [comment]
What is your model_path and tokenizer_path? One reason I can think of is that the 34B LLava uses Yi-Chat as the language model so the image token index is 64002, but vicuna-based language models will have image token index of 32000, thus causing the mismatch.

**Comment by merrymercy** [comment]
@loveunk llava-1.6-34B is not the same as the Yi-VL in this PR, although they used the same base model Yi-34B.

If you use `sgl.user/sgl. assistant`, our current chat templates only correctly handle Yi-VL and Llava-1.5.
For Llava 1.6, we need some additional handling. We can fix it soon or you can help us fix it. The related code is https://github.com/sgl-project/sglang/blob/ee1df26a77971f004444c118eabb7998dbafb14a/python/sglang/lang/chat_template.py#L171-L172.

For now, you can follow https://github.com/haotian-liu/LLaVA?tab=readme-ov-file#Demo to use llava 34B with SGLang. They handled chat template correctly in their interface.

**Comment by Lzhang-hub** [comment]
sglang=0.1.12 vllm=0.3.0

`python3 srt_example_yi_vl.py` with yi-vl-34B and `tp_size=2`
get error:
```
Traceback (most recent call last):
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/rpyc/core/protocol.py", line 369, in _dispatch_request
    res = self._HANDLERS[handler](self, *args)
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/rpyc/core/protocol.py", line 863, in _handle_call
    return obj(*args, **dict(kwargs))
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/sglang/srt/managers/router/model_rpc.py", line 62, in exposed_init_model
    self.model_runner = ModelRunner(
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/sglang/srt/managers/router/model_runner.py", line 275, in __init__
    self.load_model()
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/sglang/srt/managers/router/model_runner.py", line 308, in load_model
    model.load_weights(
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/sglang/srt/models/yivl.py", line 85, in load_weights
    self.language_model.load_weights(
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/sglang/srt/models/llama2.py", line 320, in load_weights
    weight_loader(param, loaded_weight)
  File "/home/work/l20/envs/sglang-env/lib/python3.10/site-packages/vllm/model_executor/layers/vocab_parallel_embedding.py", line 89, in weight_loader
    assert loaded_weight.shape[parallel_dim] == self.org_vocab_size
AssertionError
```

**Pull request merged**